File: app/src/main/java/com/checkin/app/checkin/Menu/UserMenu/Adapter/MenuCartAdapter.java
package com.checkin.app.checkin.Menu.UserMenu.Adapter;

import android.text.Editable;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.TextView;

import com.checkin.app.checkin.Menu.Model.OrderedItemModel;
import com.checkin.app.checkin.R;
import com.checkin.app.checkin.Utility.Utils;

import java.util.ArrayList;
import java.util.List;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import butterknife.BindView;
import butterknife.ButterKnife;

public class MenuCartAdapter extends RecyclerView.Adapter<MenuCartAdapter.ViewHolder> {
    private List<OrderedItemModel> orderedItems;
    private MenuCartInteraction mListener;

    public MenuCartAdapter(MenuCartInteraction cartInteractionListener) {
        orderedItems = new ArrayList<>();
        mListener = cartInteractionListener;
    }

    @Override
    public int getItemCount() {
        return orderedItems != null ? orderedItems.size() : 0;
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        final View itemView = LayoutInflater.from(parent.getContext()).inflate(viewType, parent, false);
        return new ViewHolder(itemView);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        holder.bindData(orderedItems.get(position));
        holder.etSpecialInstructions.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
            }

            @Override
            public void afterTextChanged(Editable s) {
                if(s.length()>0){
//                        mListener.onOrderedItemRemark(mItem, etSpecialInstructions.getText().toString());
                    orderedItems.get(position).setRemarks(holder.etSpecialInstructions.getText().toString());
                }
            }
        });
    }

    @Override
    public int getItemViewType(final int position) {
        return R.layout.item_as_menu_cart;
    }

    public void setOrderedItems(List<OrderedItemModel> orderedItems) {
        this.orderedItems = orderedItems;
//        notifyDataSetChanged();
    }

    public interface MenuCartInteraction {
        void onOrderedItemRemark(OrderedItemModel item, String s);

        void onOrderedItemChanged(OrderedItemModel item, int count);
    }

    class ViewHolder extends RecyclerView.ViewHolder {
        @BindView(R.id.tv_menu_cart_item_name)
        TextView tvItemName;
        @BindView(R.id.tv_menu_cart_item_price)
        TextView tvItemPrice;
        @BindView(R.id.tv_menu_cart_item_customized)
        TextView tvItemCustomized;
        @BindView(R.id.im_menu_cart_item_type)
        ImageView imType;
        @BindView(R.id.tv_menu_item_quantity_decrement)
        TextView tvQuantityDecrement;
        @BindView(R.id.tv_menu_item_quantity_number)
        TextView tvQuantityNumber;
        @BindView(R.id.tv_menu_item_quantity_increment)
        TextView tvQuantityIncrement;
        @BindView(R.id.et_as_menu_cart_special_instruction)
        EditText etSpecialInstructions;

        private OrderedItemModel mItem;
        private int mCount;

        ViewHolder(View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);

            tvQuantityDecrement.setOnClickListener(v -> {
                decreaseQuantity();
            });

            tvQuantityIncrement.setOnClickListener(v -> {
                increaseQuantity();
            });

            tvQuantityNumber.addTextChangedListener(new TextWatcher() {
                @Override
                public void beforeTextChanged(CharSequence s, int start, int count, int after) {
                }

                @Override
                public void onTextChanged(CharSequence s, int start, int before, int count) {
                }

                @Override
                public void afterTextChanged(Editable s) {
                        mCount = Integer.parseInt(s.toString());
                    mListener.onOrderedItemChanged(mItem, Integer.parseInt(s.toString()));

                }
            });
        }

        void bindData(final OrderedItemModel item) {
            try {
                mItem = item.clone();
            } catch (CloneNotSupportedException e) {
                e.printStackTrace();
            }

            tvItemName.setText(item.getItemModel().getName());
            tvItemPrice.setText(Utils.formatCurrencyAmount(tvItemPrice.getContext(), item.getCost()));
            if(item.isCustomized())
                tvItemCustomized.setVisibility(View.VISIBLE);
            else
                tvItemCustomized.setVisibility(View.GONE);


            if(item.getItemModel().isVegetarian())
                imType.setImageDrawable(imType.getContext().getResources().getDrawable(R.drawable.ic_veg));
            else
                imType.setImageDrawable(imType.getContext().getResources().getDrawable(R.drawable.ic_non_veg));

            displayQuantity(item.getQuantity());
            mCount = item.getQuantity();
        }

        public void setItemType(boolean isVegetarian){

        }

        /*private void setQuantityChangeListener(){

        }


        private void setOnSpecialInstructionsListener(){

        }*/

        public void increaseQuantity() {
            mCount ++;
            displayQuantity(mCount);

        }
        public void decreaseQuantity() {
            if(mCount<=0)
                return;
            mCount --;
            displayQuantity(mCount);

        }

        private void displayQuantity(int number) {
            tvQuantityNumber.setText("" + number);
        }
    }
}

File: app/src/main/java/com/checkin/app/checkin/Menu/UserMenu/Adapter/MenuCartAdapter.kt
package com.checkin.app.checkin.Menu.UserMenu.Adapter

import android.os.Handler
import android.text.Editable
import android.text.TextWatcher
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.EditText
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import butterknife.BindView
import butterknife.ButterKnife
import com.checkin.app.checkin.Menu.Model.OrderedItemModel
import com.checkin.app.checkin.R
import com.checkin.app.checkin.Utility.Utils

class MenuCartAdapter(private val mListener: MenuCartInteraction) : RecyclerView.Adapter<MenuCartAdapter.ViewHolder>() {
    private var orderedItems: List<OrderedItemModel>? = null

    override fun getItemCount(): Int = orderedItems?.size ?: 0

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder = LayoutInflater.from(parent.context).inflate(viewType, parent, false).run { ViewHolder(this) }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) = holder.bindData(orderedItems!![position])

    override fun getItemViewType(position: Int): Int = R.layout.item_as_menu_cart

    fun setOrderedItems(orderedItems: List<OrderedItemModel>) {
        this.orderedItems = orderedItems
        Handler().post { notifyDataSetChanged() }
    }

    interface MenuCartInteraction {
        fun onOrderedItemRemark(item: OrderedItemModel, s: String)

        fun onOrderedItemChanged(item: OrderedItemModel?, count: Int)
    }

    inner class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        @BindView(R.id.tv_menu_cart_item_name)
        internal lateinit var tvItemName: TextView
        @BindView(R.id.tv_menu_cart_item_price)
        internal lateinit var tvItemPrice: TextView
        @BindView(R.id.tv_menu_cart_item_customized)
        internal lateinit var tvItemCustomized: TextView
        @BindView(R.id.tv_menu_item_quantity_decrement)
        internal lateinit var tvQuantityDecrement: TextView
        @BindView(R.id.tv_menu_item_quantity_number)
        internal lateinit var tvQuantityNumber: TextView
        @BindView(R.id.tv_menu_item_quantity_increment)
        internal lateinit var tvQuantityIncrement: TextView
        @BindView(R.id.et_as_menu_cart_special_instruction)
        internal lateinit var etSpecialInstructions: EditText

        private var mItem: OrderedItemModel? = null

        init {
            ButterKnife.bind(this, itemView)

            tvQuantityDecrement.setOnClickListener { mItem?.let { mListener.onOrderedItemChanged(it, it.quantity - 1) } }

            tvQuantityIncrement.setOnClickListener { mItem?.let { mListener.onOrderedItemChanged(it, it.quantity + 1) } }

            etSpecialInstructions.addTextChangedListener(object : TextWatcher {
                override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}

                override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {}

                override fun afterTextChanged(s: Editable) {
                    mListener.onOrderedItemRemark(mItem!!, s.toString())
                }
            })
        }

        fun bindData(item: OrderedItemModel) {
            try {
                mItem = item.clone()
            } catch (e: CloneNotSupportedException) {
            }

            tvItemName.text = item.itemModel.name
            tvItemPrice.text = Utils.formatIntegralCurrencyAmount(tvItemPrice.context, item.cost)
            if (item.isCustomized) tvItemCustomized.visibility = View.VISIBLE
            else tvItemCustomized.visibility = View.GONE

            if (item.itemModel.isVegetarian) tvItemName.setCompoundDrawablesRelativeWithIntrinsicBounds(R.drawable.ic_veg, 0, 0, 0)
            else tvItemName.setCompoundDrawablesRelativeWithIntrinsicBounds(R.drawable.ic_non_veg, 0, 0, 0)

            tvQuantityNumber.text = item.quantity.toString()
        }
    }
}
